### PR TITLE
Adds a interval_list row merger tool.

### PR DESCRIPTION
**Pull request by vruano**
Can handle .tsv and .tsv.gz input and output files. 
In theory NIO ready. 
@Header lines are preserved.
No test added so far but you should be able to test it out on a given file(s) or bucket url(s) like so:

`./gatk CondenseReadCounts --input input.100bp.tsv -O output.1000bp.tsv.gz -f 10 --out-bin-length 1000`

The last argument ```--out-bin-length 1000``` is optional and make sure that only those bins with exactly 1000 bases are output.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@CommandLineProgramProperties(summary = "condenses read-counts in .tsv format into a new file also in .tsv format",
+        oneLineSummary = "condenses read-counts in .tsv format into a new file also in .tsv formas" ,
+        programGroup = CopyNumberProgramGroup.class)
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: If this needs to be merged quickly, we should add the `@ExperimentalFeature`.  This tool doesn't have documentation, lacks tests, breaks the design of the CNV collection classes, and is otherwise idiosyncratic.  It is also written primarily to satisfy a specific use case in a "downstream" pipeline.  We can go back and address the former, but should also be sure the latter is justified (@mwalker174 recall that we just discussed a related issue).  That said, I do think a tool that combines counts will be useful, so we should take care with the final design.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+public class CondenseReadCounts extends CommandLineProgram {
+
+    public static final String FACTOR_LONG_NAME = "factor";
+    public static final String FACTOR_SHORT_NAME = "f";
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I'm also not sure "factor" is the most evocative name...any other candidates?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    public static final String EXPECTED_BIN_LENGTH_LONG_NAME = "out-bin-length";
+
+    @Argument(
+            doc = "Path to the input counts file",
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: ...file in TSV format.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            fullName = StandardArgumentDefinitions.INPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME
+    )
+    private String input;
+
+    @Argument(
+            doc = "Path to the output counts file",
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: CNV argument docs end with periods.  Should fix below as well.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    @Argument(
+            doc = "Indicates how many input read count record must be condense into each output record",
+            fullName = FACTOR_LONG_NAME,
+            shortName = FACTOR_SHORT_NAME,
+            minValue = 1
+    )
+    private int factor = 0;
+
+    @Argument(
+            doc = "expected bin-length after condensation, those output bin that do not match this length will be excluded. " +
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Capitalization and typos in the doc string.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    )
+    private String output;
+
+    @Argument(
+            doc = "Indicates how many input read count record must be condense into each output record",
+            fullName = FACTOR_LONG_NAME,
+            shortName = FACTOR_SHORT_NAME,
+            minValue = 1
+    )
+    private int factor = 0;
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Why is this set to 0 by default?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    private SimpleCount composeRecord(final String contig, final int start, final int end, final int count) {
+        return new SimpleCount(new SimpleInterval(contig, start, end), count);
+    }
+
+    private class SimpleCountReader extends TableReader<SimpleCount> {
+        private SimpleCountReader(final String input) throws IOException {
+            super(input, new InputStreamReader(BucketUtils.openFile(input)));
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Enabling read/write to buckets and bypassing creation of the list of objects in the CNV collection classes seem like two separate issues that should be solved by modifying the parent classes.  Probably only the second issue really affects cost/memory issues, since the files are not so big that localization/delocalization adds any overhead, correct?

I'm fine with doing this only for TSV count files to start, but other files could benefit.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java [hunk 8]
```diff
@@ -27,16 +27,16 @@
     /**
      * CONTIG, START, END, COUNT
      */
-    enum SimpleCountTableColumn {
+    public enum SimpleCountTableColumn {
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java]: Actually, I think it'll be easier just to duplicate these privately in the tool for the time being.  Otherwise, we have to remember to go back and fix permissions.

File: src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java [hunk 9]
```diff
@@ -365,8 +365,7 @@ private R fetchNextRecord() throws IOException {
     }
 
     private void processCommentLine(final String[] line, final long lineNumber) {
-        final StringBuilder builder = new StringBuilder();
-        builder.append(line[0].substring(TableUtils.COMMENT_PREFIX.length()));
+        final StringBuilder builder = new StringBuilder(line[0]);
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java]: Did you mean `final StringBuilder builder = new StringBuilder(line[0].substring(TableUtils.COMMENT_PREFIX.length()));`?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            optional = true
+    )
+    private int expectedBinLength = 0;
+
+    @Override
+    public String doWork() {
+
+        try (final SimpleCountReader reader = new SimpleCountReader(input);
+             final SimpleCountWriter writer = new SimpleCountWriter(output)) {
+            String contig = null;
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: This logic needs tests and should be extracted.  A precise description of the method would be nice.

That said, I think some thought might be needed to nail down the method we want, just to make sure filtering/blacklisting doesn't trip us up.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+            }
+            skippedRecords += currentRecordCount;
+            skippedReadCounts += currentReadCount;
+            logger.info("Number of output records: " + condensedRecords / factor);
+            logger.info("Number of condensed input records: " + condensedRecords);
+            logger.info("Number of skipped input records: " + skippedRecords);
+            logger.info("Total count condensed: " + outputCounts);
+            logger.info("Total count skipped: " + skippedReadCounts);
+            logger.info(String.format("%% of lossed records and counts are: %.2g and %.2g",
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Why "lossed" here and "skipped" above?

Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I would also just break this message up into two lines to be consistent with the above.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    private class SimpleCountReader extends TableReader<SimpleCount> {
+        private SimpleCountReader(final String input) throws IOException {
+            super(input, new InputStreamReader(BucketUtils.openFile(input)));
+        }
+
+        private List<String> headerLines;
+        private boolean foundHeader = false;
+
+        @Override
+        public boolean isCommentLine(String[] line) {
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I think assuming this format would have the dual benefit of simplifying this code and continuing to enforce the format.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+            if (IOUtil.hasBlockCompressedExtension(output)) {
+                return new OutputStreamWriter(new BlockCompressedOutputStream(BucketUtils.createFile(output), (File) null));
+            } else {
+                return new OutputStreamWriter(BucketUtils.createFile(output));
+            }
+        }
+
+        public void writeHeaderComments(final List<String> comments) throws IOException {
+            for (final String line : comments) {
+                writeLineBypassingCSVWriter(line);
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Why is this necessary?

File: src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java (hunk 9 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java]: This is causing tests to fail.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 7 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Actually, I'm not even convinced we need to work too hard to enable streaming in this tool, since it won't be long running or memory intensive.  I'd rather pay the localization/delocalization cost than the code cost.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.List;
+
+@CommandLineProgramProperties(summary = "condenses read-counts in .tsv format into a new file also in .tsv format",
+        oneLineSummary = "condenses read-counts in .tsv format into a new file also in .tsv formas" ,
+        programGroup = CopyNumberProgramGroup.class)
+public class CondenseReadCounts extends CommandLineProgram {
+
+    public static final String FACTOR_LONG_NAME = "factor";
+    public static final String FACTOR_SHORT_NAME = "f";
+    public static final String EXPECTED_BIN_LENGTH_LONG_NAME = "out-bin-length";
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Or maybe something including "filter" in the name.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@CommandLineProgramProperties(summary = "condenses read-counts in .tsv format into a new file also in .tsv format",
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I would suggest "combine" or "coalesce" rather than "condense"

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: As a suggestion, "num-bins" seems intuitive.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 11 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: lossed -> lost

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 1 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: formas -> format

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+    @Argument(
+            doc = "Path to the output counts file",
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME
+    )
+    private String output;
+
+    @Argument(
+            doc = "Indicates how many input read count record must be condense into each output record",
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I'd suggest `Indicates how many consecutive input read count records will be combined for each output record`

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+            fullName = FACTOR_LONG_NAME,
+            shortName = FACTOR_SHORT_NAME,
+            minValue = 1
+    )
+    private int factor = 0;
+
+    @Argument(
+            doc = "expected bin-length after condensation, those output bin that do not match this length will be excluded. " +
+                    "The default 0 means that such constrain is not be applied",
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: `such constrain` -> `this constraint`

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 7 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I agree, let's use `SimpleCountCollection` to handle the counts. We can go back and modify those for streaming performance later.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java]: Why the column composition of the count files should be a private matter? Perhaps I could move my class the enclosing package to be able to access. Duplicate a enum and make sure they keep consistent is not a good idea.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: No problem in removing the "#".

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: So no -f, that is --fine :) 
"factor" as in "condesnation factor" --factor 10  means 10 times condensed. If you insist it can be "num-bins" though.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: --emit-only-bin-of-length 1000?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SimpleCountCollection.java]: "Duplicate...for the time being"!  Easier to fix the idiosyncratic code here later, rather than remember what we changed in the CNV package.

The file is a 1:1 representation of the Java class and any code that needs to talk to it should be encapsulated in this package, if not the class itself.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: "number-of-bins" is even better.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: In any case this is a bit of a hack in my opinion .... perhaps it should be a range or min... considering that is totally possible for the input to have a mix of bins sizes... in the problem at hand  some bins have size ~950bp .... which sounds acceptable for output. I added this argument in case that we decide to exclude those we can do it quickly.

[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: I think verb names should be left to booleans.  Perhaps "emitted-bin-length"?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: So you want to change the name of the tool to Combine?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 6 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Yeah, is what Java does anyway so has no effect. In any case is not optional so that default is irrelevant.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: You might recall that we used to have a CombineReadCounts tool, which did a completely different thing...

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 7 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Streaming does not make the code more complicate ... the more complex code (the conderser loop) would still apply with a fully load collection. 

About NIO ... BucketUtils make it kinda transparent whether it is actual NIO or a local file... basically we change "BuckedUtils.openFile" by  "new FileInputStream(". Also has the added benefit of handling ".gzs"

File: src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java (hunk 9 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/utils/tsv/TableReader.java]: Fixed the bug... The comment starting character (in fact the first character of the comment line) was stripped so I was loosing the "@". This only affects another place and went it and changed it...

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 10 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Will try to make the code a bit more readable and add documentation... but it is quite simple what it does... it is the more simple implementation of condensing adjacent counts i.e. greedily from up-to-downstream. Basically its documentation would be in the description of the tool.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 13 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Is to avoid "forced" quotation by CSVWriter of the @ header line ... that said ,looking deeper in CSVWriter code I might be able to use this backdoor.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 7 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: You only need this custom SimpleCountReader so you can stream, correct?

It's great that BucketUtils makes it easy to enable streaming, so maybe we can enable this at the TableWriter level where it would have the most benefit.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 10 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: If the Javadoc is sufficiently precise, this is fine.  Or you can have more user friendly and breezy Javadoc and leave the details to comments here or the backend method documentation.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Also, specifying a range might be good.  You could imagine getting gigantic bins, depending on your blacklist.  This could be a problem if ever the model wants to use bin length as a covariate.  Let's discuss for the final tool design.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 7 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Also, just sticking to the vanilla SimpleCountCollection read/write allows the tool to operate on HDF5.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: Perhaps just "RebinReadCounts" would suffice.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java (hunk 13 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CondenseReadCounts.java]: You create the underlying writer in this class and should already be able to use it without going through the TableWriter.  We do a similar thing in AbstractRecordCollection to write the SAM header.
